### ISSUE TITLE
Should the R package also install the headers?

### ISSUE THREAD

**Issue by eddelbuettel** [mlpack/mlpack#3486]
### What is the desired addition or change?

I was answering a [StackOverflow question about using mlpack from R](https://stackoverflow.com/questions/76319284/problem-with-using-rcppmlpack-in-my-own-r-package) and it became apparent that we do install the headers (see detail below).  That limiits the use of the package somewhat.

<details>

Ubuntu 22.10, all CRAN packages current, `mlpack` from CRAN.

```sh
edd@rob:~$ tree /usr/local/lib/R/site-library/mlpack/
/usr/local/lib/R/site-library/mlpack/
├── CITATION
├── DESCRIPTION
├── help
│   ├── aliases.rds
│   ├── AnIndex
│   ├── mlpack.rdb
│   ├── mlpack.rdx
│   └── paths.rds
├── html
│   ├── 00Index.html
│   └── R.css
├── include
│   └── mlpack.h
├── INDEX
├── libs
│   └── mlpack.so
├── LICENSE
├── Meta
│   ├── features.rds
│   ├── hsearch.rds
│   ├── links.rds
│   ├── nsInfo.rds
│   ├── package.rds
│   └── Rd.rds
├── NAMESPACE
└── R
    ├── mlpack
    ├── mlpack.rdb
    └── mlpack.rdx

6 directories, 23 files
edd@rob:~$ 
```

</details>

### What is the motivation for this feature?

Making the package more useable

### If applicable, describe how this feature would be implemented.

Modify `CMakeLists.txt` to coinstall what goes to, say, `/usr/include/mlpack/` to also go to `inst/include/` which becomes `include/` in the R package (and which R points back to via the common `LinkingTo: mlpack` a package user would use.

### Additional information?

N/A

**Comment by eddelbuettel** [comment]
Yup and yup!!  (I 'lazily' caught up yesterday and used my Launchpad PPA to build an Ubuntu binary of Debian's most recent package, mlpack 4.0.1 (and filed a Debian bug report to kindly roll forward to last month's 4.1.0) and verified that yes indeed we only need the headers).

**Bulletpoint one:**  Our good friends RcppArmadillo, and then of course also RcppEnsmallen, are prime example.  See how they use the directory `inst/include/`:  [RcppArmadillo](https://github.com/RcppCore/RcppArmadillo/tree/master/inst/include) which contains Conrad's code unchanged (I just the unused cmake config setter as no cmake is touched in an R packahe build) and [RcppEnsmallen](https://github.com/coatless-rpkg/rcppensmallen/tree/master/inst/include) doess the exact same thing.  In fact, so do _hundreds_ of R package.  The key that _every directory or file below `inst/`_ ends up being installed 'as-is' in an R package installation.  So is the perfect vessel for header librariess -- and I have done just that for 10+ years for Armadillo, a subset of Boost (package BH), Eigen, Asio, ... and more.   

R uses this by declaring `LinkingTo: ` in the DESCRIPTION file.  That tells R to construct a `-I....` flag to whereever the package that is referenced on that machine points.  And ... that is all.   Eg when I install an old user of RcppArmadillo I see (to grab just one line)

     ccache g++-12 -I"/usr/share/R/include" -DNDEBUG  \
               -I'/usr/local/lib/R/site-library/Rcpp/include' \
               -I'/usr/local/lib/R/site-library/RcppArmadillo/include'   \
               -fpic  -g -O3 -Wall -pipe -pedantic   -c expmMat.cpp -o expmMat.o

because the package in question has 

      LinkingTo: Rcpp, RcppArmadillo

So users of (R package) `mlpack` would just need to add `LinkingTo: mlpack`  (well:  and RcppArmadillo, RcppEnsmallen) and be done.   

In an ad-hoc R session you that taken care of by `Rcpp::sourceCpp()` because we declare `// [[Rcpp::depends(RcppArmadillo)]]` which does the same thing in non-package builds: Add the needed -I....

This _just works_ and is fabulous.   Case in point is my [updated SO answer](https://stackoverflow.com/a/76333026/143305) showing use of RcppArmadillo, RcppEnsmallen, mlpack (header-less, relies on my system headers, this would fix it) and a switch to C++14.

**Bulletpoint two:**  My `cmake`-foo is weak but just how we in recent PRs for the R package added an explicit copy of a single file here or there we would have to ensure that the headers end up in `inst/include`  as

      inst/include/mlpack.hpp
      inst/include/mlpack/**             # recursively

and then we should be done.   That is probably handled either or two install statements and I would hope we do not have to enumerate all the include header files explicitly.

**Comment by rcurtin** [comment]
I see!  Thanks for the comprehensive answer.  If everything in `inst/` is installed automatically, then I would just need to move the mlpack headers from `src/` to `inst/`, and maybe make some small updates so that the package still builds correctly (easy enough).

However, one little gotcha: we don't use the `RcppCereal` package because it throws warnings on CRAN.  Here's the relevant comment in the `CMakeLists.txt`:

```
  # In case of R_Bindings we are copying all of these header to
  # <package>/src/cereal/ because the currently-packaged version
  # of Rcereal throws warnings on CRAN.
```

So, should I also put the cereal headers into `inst/`?  Or could we run into issues where a user has installed `Rcereal`, and then mlpack attempts to install new headers over it?  Without the cereal headers, compilation will fail.

**Comment by eddelbuettel** [comment]
I see -- it didn't fail for me here but maybe I 'got lucky' as I have libcereal-dev 1.3.2 installed.  I could try in a container to use a more restricted cleaner setup.

If `cereal` is needed, it should go to `inst/include` too and we would effectively 'vendor' it.  (And one of the two of use should maybe talk to Wush to update `Rcereal` but that is for another day).

**Comment by rcurtin** [comment]
Awesome.  Thanks again for the help and explanation.  I still want to see it build on Azure Pipelines, but I'm pretty sure #3488 handles this. :+1:

**Comment by conradsnicta** [comment]
resolved in #3488

**Comment by eddelbuettel** [comment]
And on CRAN -- love to see it.  https://mastodon.social/@CRANberriesFeed@mas.to/110605692538355162 ie

![image](https://github.com/mlpack/mlpack/assets/673121/63875f98-f0e4-49ac-8b64-45a7bbf26c60)

and 

![image](https://github.com/mlpack/mlpack/assets/673121/f3c4dbe9-7dc0-4d55-9e77-cb67eea13ab2)
